test/parallel/test-async-wrap-asyncresource-constructor.js

```diff
@@ -0,0 +1,29 @@
+'use strict';
+require('../common');
+
+// This tests that AsyncResource throws an error if bad parameters are passed
+
+const assert = require('assert');
+const AsyncResource = require('async_hooks').AsyncResource;
+const async_wrap = process.binding('async_wrap');
+
+/* without this, there is nothing to do,
+* and the constructor would return immediately
+*/
+async_wrap.async_hook_fields[async_wrap.constants.kInit] = 1;
```

Comment:
Refering to https://github.com/nodejs/node/pull/13336#discussion_r119815621 the logic there in `AsyncResource()` is bad, so this shouldn't be necessary. If this lands first we need a comment or such that this should be removed.

Also, I'd rather see
```js
async_hooks.createHook({ init() { } }).enable();
```
than manipulating `kInit` manually.